carmajava/signal_plugin/src/main/java/gov/dot/fhwa/saxton/carma/signal_plugin/TrafficSignalPlugin.java

```diff
@@ -634,33 +678,82 @@ public void loop() throws InterruptedException {
-            if (awaitingUserConfirmation.compareAndSet(false, true)) {
-                // Update state variables
-                prevUIRequestTime.set(System.currentTimeMillis());
-                awaitingUserInput.set(true);
-                // Send user input request to get confirmation to continue
-                askUserIfIntersectionIsClear();
-                log.info("Requesting user confirmation");
 
-            } else if (System.currentTimeMillis() - UI_REQUEST_INTERVAL > prevUIRequestTime.get() // Already awaiting confirmation and enough time has elapsed
-                && awaitingUserInput.compareAndSet(false, true)) { // AND we are not currently waiting for user input
-                // Then we want to request input from the user again
-                askUserIfIntersectionIsClear();
-                log.info("Re-Requesting user confirmation");
+        log.info("evaluateStatesForStopping 0a: stoppedForLight: " + stoppedForLight  + "; phaseIsGreenSoon: " + phaseIsGreenSoon + "; phaseIsGreen: " + phaseIsGreen);
+        log.info("evaluateStatesForStopping 0b: awaitingUserConfirmation: " + awaitingUserConfirmation + "; awaitingUserInput:" + awaitingUserInput + "; userRequestedToWait: " + userRequestedToWait);
+
+        //The userRequestedToWait is created to track the GO_BUTTON_SRV response, since the thread here needs to stop.
+        if (userRequestedToWait.compareAndSet(true, false)){
+            try {
+                log.info("evaluateStatesForStopping: BEFORE sleep");
+                Thread.sleep(WAIT_INTERVAL);
+                log.info("evaluateStatesForStopping: AFTER sleep");
+            } catch(InterruptedException e) {
+                Thread.currentThread().interrupt();
+                log.info("evaluateStatesForStopping: Failed TO SLEEP");
+                //do not throw exception
             }
-        } else if (!stoppedForLight && awaitingUserConfirmation.compareAndSet(true, false)) {
+        }
+
+        // If stopped & light is green/green soon, & have NOT requested user input/confirmation
+        if (stoppedForLight
+          && phaseIsGreenSoon
+          && awaitingUserConfirmation.compareAndSet(false, true)
+          && awaitingUserInput.compareAndSet(false, true))
+        {
+            // Send user input request to get confirmation to continue
+            askUserIfIntersectionIsClear();
+
+            log.info("evaluateStatesForStopping 1: New Request: awaitingUserConfirmation: " + awaitingUserConfirmation + "; awaitingUserInput:" + awaitingUserInput);
+        }
+        // If stopped & light is green/green soon, & still waiting for YES confirmation but NOT user input, & enough time has elapsed
+        else if (stoppedForLight
+          && phaseIsGreenSoon
+          && awaitingUserConfirmation.get()
+          && awaitingUserInput.compareAndSet(false, true)
+          && System.currentTimeMillis() - UI_REQUEST_INTERVAL > prevUIRequestTime.get())
+        {
```

Comment:
I'm a bit unclear on what this time check is for compared to the wait time above. It seems like they are both geared towards accomplishing the same thing